### PR TITLE
Add priority flag to distributions

### PR DESCRIPTION
**Pull request by hschallhorn**
Bumps #14604

### Description
Allows us to signify a distribution includes only priority cases pushed to a judge. Will allow us to filter by priority when calculating fairness for number of cases pushed to judges. Also allows us to skip some validations that we do not care about when pushing cases to judges.

### Acceptance Criteria
- [ ] Distributions can be created with a priority flag
- [ ] Validation for too_many_unassigned_cases is skipped when priority is true
- [ ] Validation for unassigned_cases_waiting_too_long is skipped when priority is true
- [ ] Validation for pending_distribution only checks for pending distributions of the same priority (a judge can have a pending distribution that they requested while still receiving a push priority case

### Database Changes
*Only for Schema Changes*

* [x] Column comments updated
* [x] Query profiling performed (eyeball Rails log, check bullet and fasterer output)
* [x] DB schema docs updated with `make docs`
* [x] #appeals-schema notified with summary and link to this PR

### CODE REVIEW

File: docs/schema/caseflow.csv [hunk 1]
```diff
@@ -687,18 +688,18 @@ legacy_issues,vacols_sequence_id,integer ∗,x,,,,,The sequence ID of the lega
...
+legacy_issue_optins,original_legacy_appeal_disposition_code,string,,,,,,The original disposition code of legacy appeal being opted in
 legacy_issue_optins,request_issue_id,integer (8) ∗ FK,x,,x,,x,The request issue connected to the legacy VACOLS issue that has been opted in.
 legacy_issue_optins,rollback_created_at,datetime,,,,,,"Timestamp for when the connected request issue is removed from a Decision Review during edit, indicating that the opt-in needs to be rolled back."
 legacy_issue_optins,rollback_processed_at,datetime,,,,,,Timestamp for when a rolled back opt-in has successfully finished being rolled back.
 legacy_issue_optins,updated_at,datetime ∗,x,,,,x,Automatically populated when the record is updated.
-legacy_issue_optins,original_legacy_appeal_decision_date,datetime,,,,,,The original disposition date of a legacy appeal being opted in.
-legacy_issue_optins,original_legacy_appeal_disposition_code,string,,,,,,The original disposition code of legacy appeal being opted in.
-legacy_issue_optins,folder_decision_date,date,,,,,,Decision date on case record folder.
```
Comment (hschallhorn) [inline docs/schema/caseflow.csv:701]: Results of running make docs

File: app/controllers/distributions_controller.rb [hunk 2]
```diff
@@ -102,7 +102,7 @@ def json_error(error)
   end
 
   def pending_distribution
-    Distribution.pending_for_judge(judge)
+    Distribution.pending_for_judge(judge).first
```
Comment (yoomlam) [inline app/controllers/distributions_controller.rb:105]: Suggest renaming method to `first_pending_distribution` to imply that there can be more than one.
Alternative is to change the method to: 
```ruby
  def pending_distributions
    Distribution.pending_for_judge(judge)
```
and let the caller decide which distribution it wants to use.

File: db/migrate/20200713173228_add_priority_to_distributions.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class AddPriorityToDistributions < ActiveRecord::Migration[5.2]
+  def up
+    add_column :distributions, :priority, :boolean, comment: "Whether or not this distribution was a priority appeals only push to judges via weekly job (not requested)"
```
Comment (yoomlam) [inline db/migrate/20200713173228_add_priority_to_distributions.rb:3]:
```suggestion
    add_column :distributions, :priority_pushed, :boolean, comment: "Whether or not this distribution was a result of a priority-appeals-only push to judges via a weekly job (not requested)"
```

File: db/migrate/20200713173228_add_priority_to_distributions.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+    change_column_default :distributions, :priority, false
+  end
+
+  def down
+    remove_column :distributions, :priority
+  end
+end
```
Comment (yoomlam) [inline db/migrate/20200713173228_add_priority_to_distributions.rb:10]: First time I've seen this symbol for "No newline at end of file". I recently read https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline, so I suggest adding a newline.

File: db/migrate/20200713174152_backfill_distribution_priority.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class BackfillDistributionPriority < ActiveRecord::Migration[5.2]
+  disable_ddl_transaction!
```
Comment (yoomlam) [inline db/migrate/20200713174152_backfill_distribution_priority.rb:2]: TIL: https://stackoverflow.com/questions/46066810/when-would-you-want-to-use-disable-ddl-transaction-in-rails

So will we be notified if it fails to update a record? Or do we just check prod after this PR is deployed?

File: db/schema.rb [hunk 6]
```diff
@@ -843,7 +844,7 @@
     t.boolean "military_sexual_trauma", default: false, comment: "Military Sexual Trauma (MST)"
     t.boolean "mustard_gas", default: false
     t.boolean "national_cemetery_administration", default: false
-    t.boolean "no_special_issues", default: false, comment: "Affirmative no special issues; column added belatedly"
+    t.boolean "no_special_issues", default: false, comment: "Affirmative no special issues, added belatedly"
```
Comment (yoomlam) [inline db/schema.rb:847]:
```suggestion
    t.boolean "no_special_issues", default: false, comment: "Affirmative no special issues; added belatedly"
```

File: db/migrate/20200713173228_add_priority_to_distributions.rb (hunk 3 above)
Comment (yoomlam) [reply db/migrate/20200713173228_add_priority_to_distributions.rb:3]: Re-reading this. Maybe this is more appropriate:
```suggestion
    add_column :distributions, :priority, :boolean, comment: "Whether or not this distribution is a priority-appeals-only push to judges via a weekly job (not manually requested)"
```

File: db/migrate/20200713174152_backfill_distribution_priority.rb (hunk 5 above)
Comment (hschallhorn) [reply db/migrate/20200713174152_backfill_distribution_priority.rb:2]: As unsafe migration was caught by our [strong_migrations](https://github.com/ankane/strong_migrations) gem and it provides this as the solution.

output
```
StandardError: An error has occurred, this and all later migrations canceled:

=== Dangerous operation detected #strong_migrations ===

Adding a column with a non-null default causes the entire table to be rewritten.
Instead, add the column without a default value, then change the default.

class StrongMigrations::Checker < ActiveRecord::Migration[5.2]
  def up
    add_column :distributions, :priority, :boolean, comment: "Whether or not this distribution was a priority appeals only push to judges via weekly job (not requested)"
    change_column_default :distributions, :priority, false
  end

  def down
    remove_column :distributions, :priority
  end
end

Then backfill the existing rows in the Rails console or a separate migration with disable_ddl_transaction!.

class BackfillStrongMigrations::Checker < ActiveRecord::Migration[5.2]
  disable_ddl_transaction!

  def change
    Distribution.unscoped.in_batches do |relation|
      relation.update_all priority: false
      sleep(0.1)
    end
  end
end
```

File: app/controllers/distributions_controller.rb [hunk 7]
```diff
@@ -38,7 +38,7 @@ def enqueue_distribution_job(distribution)
 
   def render_errors(error)
     errors = error.record.errors.details.values.flatten.map { |error_details| error_details[:error] }
-    return render_single(pending_distribution) if errors.include? :pending_distribution
+    return render_single(pending_distributions.first) if errors.include? :pending_distribution
```
Comment (hschallhorn) [inline app/controllers/distributions_controller.rb:41]: https://github.com/department-of-veterans-affairs/caseflow/pull/14686/files#diff-6e7452c03ad40ddef41c49ff0ea1e0d8R25 now returns more than one

File: app/models/distribution.rb [hunk 8]
```diff
@@ -74,7 +74,9 @@ def validate_days_waiting_of_unassigned_cases
   end
 
   def validate_judge_has_no_pending_distributions
-    errors.add(:judge, :pending_distribution) if self.class.pending_for_judge(judge)
+    if self.class.pending_for_judge(judge).where(priority_push: priority_push).exists?
+      errors.add(:judge, :pending_distribution)
+    end
```
Comment (hschallhorn) [inline app/models/distribution.rb:79]: We only care about pending distributions of the same priority. This allows a user to request more cases, even if there is a priority push in progress for them.

**Pull request merged**